In [94]:
import numpy as np
import pandas as pd

comment_data = pd.read_csv("Sample_labaled_data.csv")

In [95]:
 from googleapiclient.discovery import build

import json


def get_toxicity_score(comment):
    
    API_KEY = 'my-api-key'
    
    client = build(
    "commentanalyzer"
    "v1alpha1",
    developerKey=API_KEY,
    discoveryServiceUrl="https://commentanalyzer.googleapis.com/$discovery/rest?version=v1alpha1",
    static_discovery=False,
    version=2.0,
    )
    
    analyze_request = {
    'comment': {'text' : comment }, 
    'requestedAttributes': {'TOXICITY': {}}
    }
    
    response = client.comments().analyze(body=analyze_request).execute()
    toxicity_score = (response["attributeScores"]["TOXICITY"]["summaryScore"]["value"])*100
    
    return toxicity_score



toxicity_scores = []
i=0
comment_data_small=comment_data.head(5000)
for index, row in comment_data_small.iterrows():
    comment = row['comment_text']
    try:
      score = get_toxicity_score(comment)
    except:
      score = 0
    toxicity_scores.append(score)
    print(i)
    i=i+1

comment_data_small['toxicity_score'] = toxicity_scores

comment_data_small.to_csv("comment_data_updated.csv")

In [96]:
comment_data=pd.read_csv("comment_data_updated.csv")
pd.set_option('max_colwidth', 5000)
comment_data.head(50)

Unnamed: 0.1  Unnamed: 0                id  \
0              0           5  0001ea8717f6de06   
1              1           7  000247e83dcc1211   
2              2          11  0002f87b16116a7f   
3              3          13  0003e1cccfd5a40a   
4              4          14  00059ace3e3e9a53   
5              5          16  000663aff0fffc80   
6              6          17  000689dd34e20979   
7              7          19  000844b52dee5f3f   
8              8          21  00091c35fa9d0465   
9              9          22  000968ce11f5ee34   
10            10          26  000a02d807ae0254   
11            11          28  000bafe2080bba82   
12            12          29  000bf0a9894b2807   
13            13          31  000c9b92318552d1   
14            14          33  000cf60dbaed8c02   
15            15          34  000d4f120d5a7303   
16            16          38  001068b809feee6b   
17            17          40  0011cefc680993ba   
18            18          41  0011ef6aa33d42e6   
19            19          43  0012bb72f20ae971   
20            20          45  00137446b1aec28c   
21            21          46  0013a435effa29bd   
22            22          47  0013be435187e84f   
23            23          49  001411adf8f1dd82   
24            24          51  001426f56de6a49b   
25            25          56  0016b94c8b20ffa6   
26            26          58  00177176f33f587e   
27            27          60  001804717513d860   
28            28          62  0019400581a6fe32   
29            29          65  001acc29b248cb6d   
30            30          69  001c368c738a814f   
31            31          72  001ceacfd4ffa469   
32            32          73  001ceebc43e65027   
33            33          75  001d39c71fce6f78   
34            34          76  001d739c97bc2ae4   
35            35          79  001eba8bbe728e54   
36            36          81  001eff4007dbb65b   
37            37          89  0020f1b79913e7c0   
38            38          93  0023317d123b6fcb   
39            39          94  00234e493cedb3e1   
40            40          95  0023f3f84f353bce   
41            41          97  0025a91b6955f1a5   
42            42          98  0025c49d87d9a18f   
43            43         100  0026c91011e3ae35   
44            44         105  0028bca06339e02f   
45            45         106  0028c5110f1f896b   
46            46         108  00293511a845d344   
47            47         110  002af5c4c7f08722   
48            48         116  002eef919d2f8410   
49            49         119  003044a2c35274b6   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             

In [97]:
threshold = comment_data.loc[(comment_data['toxicity_score'] >= 0.4) & (comment_data['toxicity_score'] <= 0.5)]
threshold.head(50)


Unnamed: 0.1  Unnamed: 0                id  \
11              11          28  000bafe2080bba82   
29              29          65  001acc29b248cb6d   
56              56         131  00372dfdd531fc6c   
69              69         184  004a5c6fc6ad7ee1   
159            159         437  00bbaedc072b0d95   
178            178         484  00cbc079f3be0661   
402            402        1068  01c6333adb815769   
544            544        1490  0281308c885e9540   
550            550        1507  0289e205824f6128   
566            566        1543  029c0c7035a1dd3c   
604            604        1644  02c9053d5b8cd9f5   
668            668        1831  032422b6e5118d59   
681            681        1863  0334dfe7e93a4184   
683            683        1865  0335d0a5ae95db1b   
695            695        1898  033fb0d5f54acaf0   
696            696        1901  0341271c4b0e1241   
727            727        2022  037b957e2f131c73   
734            734        2045  03876c134897ae1e   
805            805        2251  03e12f726b274e3b   
946            946        2669  049c0a9fc2d606ff   
950            950        2685  04a264523e42eb09   
954            954        2698  04a996318ad46b52   
965            965        2719  04b1bfcfa5223233   
995            995        2833  04e68c09efa4e861   
1042          1042        2952  051a2e6727136c0a   
1142          1142        3222  058de878deb0a50a   
1188          1188        3336  05ba7e8f954be065   
1192          1192        3344  05bec37c88d07f5c   
1356          1356        3792  0683a48f016a624a   
1387          1387        3896  06a969e6132c50f3   
1395          1395        3921  06b59f9d648b3239   
1402          1402        3946  06bde556ae32c4ce   
1409          1409        3974  06cc9bb3724b18f7   
1433          1433        4026  06e6281cfa6831f2   
1434          1434        4028  06e75aa9394c8d06   
1560          1560        4364  076c69a3d5479ec0   
1563          1563        4370  076f0d55455492f9   
1584          1584        4445  07902821157aab3f   
1601          1601        4488  07a145da57085090   
1800          1800        4987  086ae0d412d62ab7   
1806          1806        4996  086e5cb0920e7b2c   
1964          1964        5400  0919d76c09afddb3   
1984          1984        5443  092f841409dfe212   
1991          1991        5484  093ea9c88a986054   
1997          1997        5496          9.43E+21   
2000          2000        5505  09463ceb22143302   
2144          2144        5909  09f7db293985f37b   
2151          2151        5939  0a01379fb4763152   
2158          2158        5958  0a0531c2d205c838   
2165          2165        5977  0a0d0ae781f2bdd2   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         

In [98]:
threshold = comment_data.loc[(comment_data['toxicity_score'] >= 0.5) & (comment_data['toxicity_score'] <= 0.6)]
threshold.head(50)


Unnamed: 0.1  Unnamed: 0                id  \
34              34          76  001d739c97bc2ae4   
51              51         121  0030e3ba49006440   
134            134         370  009ba60f01432c26   
136            136         372  009d713f2af2ad50   
206            206         551  00e8e19baaa1891f   
327            327         880  017319eb69e7846b   
344            344         915  0185ac72a68744d8   
380            380        1011  01afcfda9587d3e4   
535            535        1468  02793b255d522ff2   
563            563        1540  029a12baa96789d7   
710            710        1981  0365e901cca2e2c1   
963            963        2717  04b186645b729105   
985            985        2805  04dbc74942771d05   
1148          1148        3235  05928f8cd3de8daf   
1154          1154        3253  059b68ba5fc6f5f6   
1180          1180        3321  05b48cab97335fb2   
1259          1259        3537  061491ecb786f049   
1315          1315        3670  064e64b06c36a263   
1414          1414        3985  06d3c3a868d29d62   
1575          1575        4408  077f2ec4b2400709   
1580          1580        4433  078b225068f10354   
1581          1581        4434  078bc2569dddebc9   
1603          1603        4497  07a4f4352c49d002   
1762          1762        4882  083a66765ac18d47   
1785          1785        4948  085ae09daf63dc75   
1786          1786        4949  085b65eae0da965d   
1797          1797        4980  0867cc556f2a744b   
1942          1942        5346  0902024acc4c6ebe   
1955          1955        5381  090f410ea580df44   
1992          1992        5486  093f933f0e8f93af   
2142          2142        5905  09f68b4ed2e6039e   
2148          2148        5926  09fd868e59e66656   
2166          2166        5981  0a0e626537b10a1f   
2171          2171        5987  0a10be8ea2de9e31   
2177          2177        5999  0a16baa76026801f   
2178          2178        6001  0a1713674693992f   
2193          2193        6041  0a294e762529f68d   
2213          2213        6105  0a43327d97fd7bde   
2480          2480        6851  0b886485c6ac12eb   
2514          2514        6929  0bafdcdb4bbc1c31   
2529          2529        6977  0bc5ab20ad39a3ae   
2793          2793        7748  0d024fd7cbd9cd28   
2803          2803        7771  0d0cc1793fb90c86   
2808          2808        7781  0d11245014c21a83   
3026          3026        8428  0e215c231eff9433   
3049          3049        8506  0e46711c57ee0662   
3069          3069        8593  0e6cd36c238eb05b   
3074          3074        8606  0e6eeb18ce578184   
3089          3089        8650  0e801c7fb348c1d6   
3330          3330        9317  0f8d255ddc024ee9   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         

### Threshhold

After analzyzing comments within the 0.4 to 0.6 range I have determined that 0.5 and above is a good threshold for what is considered a toxic comment. 



In [99]:
toxic_comments=comment_data.loc[comment_data['toxicity_score']>=0.5]
toxic_comments.head(50)

Unnamed: 0.1  Unnamed: 0                id  \
34              34          76  001d739c97bc2ae4   
36              36          81  001eff4007dbb65b   
51              51         121  0030e3ba49006440   
99              99         260  0072f0af33e59b26   
114            114         302  0081b14d79f54b31   
132            132         361  00991697be5ed76a   
134            134         370  009ba60f01432c26   
136            136         372  009d713f2af2ad50   
137            137         376  009f03897f22b7a7   
151            151         416  00b3813b966af7e8   
162            162         445  00bd66c9ef023f41   
171            171         461  00c070a4753b0fa3   
172            172         463  00c0c355b819c5c2   
196            196         533  00dfada94844220d   
206            206         551  00e8e19baaa1891f   
327            327         880  017319eb69e7846b   
344            344         915  0185ac72a68744d8   
353            353         943  01953ce07a95be75   
373            373         996  01a9ebf72533595b   
380            380        1011  01afcfda9587d3e4   
383            383        1021  01b3bcfa50af3e68   
384            384        1024  01b44755442cf8ad   
422            422        1120  01d972d33d1d9473   
525            525        1449  026d01ace14047cb   
531            531        1461  0274ee8dcaa0b25b   
535            535        1468  02793b255d522ff2   
563            563        1540  029a12baa96789d7   
567            567        1548  029cf32eda80dee9   
568            568        1549  029cfc817949fc10   
569            569        1553  029ed93c3be25a99   
571            571        1555  029f32a30724a68b   
577            577        1573  02a74f94073757e6   
605            605        1647  02cb2ece2b20993d   
675            675        1845  032ae3155cad520b   
689            689        1886  033cf25013e93b5c   
699            699        1922  034bbd9cc8de8098   
703            703        1946  035a8db73f5b69ec   
706            706        1966  03612252a3b90a25   
710            710        1981  0365e901cca2e2c1   
718            718        2003  037206d1d78c822b   
913            913        2599  0482bc14ad24a668   
932            932        2639  0492437ae05264db   
963            963        2717  04b186645b729105   
976            976        2767  04ca01169f6c1a0e   
985            985        2805  04dbc74942771d05   
1148          1148        3235  05928f8cd3de8daf   
1149          1149        3237  0592b7c48ec6a2e2   
1152          1152        3250  0598e041cdc3ea2f   
1154          1154        3253  059b68ba5fc6f5f6   
1160          1160        3267  059fea4f6324e81f   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         

## Hypothesis

Perspective API's model is more bias towards comments using direct langauge and verbage. For example comments with phrases such as "you are" or "he is" will genrally result in higher toxicity scores than comments using less or no direct language.

In [100]:
test_df=pd.DataFrame()
selected_rows = comment_data.iloc[[34,36,132,162,171,353,383,531,675,706,710, 699,1152]]
test_df=test_df.append(selected_rows)
scores=[]
for index, row in test_df.iterrows():
    score=row['toxicity_score']*100
    scores.append(score)
test_df['toxicity_score']=scores
test_df


/var/folders/9t/76xml3rn51sbly2kmtjskwmw0000gn/T/ipykernel_725/785494671.py:3: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  test_df=test_df.append(selected_rows)


Unnamed: 0.1  Unnamed: 0                id  \
34              34          76  001d739c97bc2ae4   
36              36          81  001eff4007dbb65b   
132            132         361  00991697be5ed76a   
162            162         445  00bd66c9ef023f41   
171            171         461  00c070a4753b0fa3   
353            353         943  01953ce07a95be75   
383            383        1021  01b3bcfa50af3e68   
531            531        1461  0274ee8dcaa0b25b   
675            675        1845  032ae3155cad520b   
706            706        1966  03612252a3b90a25   
710            710        1981  0365e901cca2e2c1   
699            699        1922  034bbd9cc8de8098   
1152          1152        3250  0598e041cdc3ea2f   

                                                                                                                                     comment_text  \
34                                                               How dare you vandalize that page about the HMS Beagle Dont vandalize again demon   
36                                                                                  No he is an arrogant self serving immature idiot Get it right   
132                                                                                                                         Simple You are stupid   
162         Being drunk made Gibson say what he really believes The little freak is a Nazi like his father and I hope he dies of lung cancer Soon   
171                                                                                                        Hey shithead stop vandilizing articles   
353   You wikipedia nazis need to stop removing information just because YOU dont know if a piece of information is true doesnt mean its not true   
383                                                                                                                       kevin bravo is a faggot   
531                                                                                  I still am annoyed at you blockig me and DMacks is an idiot    
675                                                                                      Why arent you banned yet you homophobic Mexican arsehole   
706                                                      Are you serious   Seriously what purpose does flagging my article for deletion serve Fag   
710                 Maybe youre just ashamed of enjoying looking at the picture and dont want it to be cp because you are ashamed of being a pedo   
699    wow So making edits to my own talk page is considered vandalism wow just wow ED has you wikipedophiles dead on You really do have no lives   
1152                                          Oh I also take 1000 responsibility I am a worthless piece of shit that should be lynched and killed   

     toxic  toxicity_score  
34     yes       57.219880  
36     yes       87.787020  
132    yes       89.882380  
162    yes       81.156270  
171    yes       86.967080  
353     no       60.254884  
383    yes       86.296720  
531    yes       84.602730  
675    yes       91.625386  
706    yes       69.899110  
710    yes       58.851710  
699     no       71.894300  
1152   yes       91.103500

In [101]:
test_2_df=pd.DataFrame({'Comment':['How dare you vandalize that page about the HMS Beagle you are a demon dont vandalize again',
                                   'No that arrogant self serving immature idiot Get it right',
                                   ' Its simple stupidity',
                                   'Being drunk made Gibson say what he really believes The little freak Nazi is like his father and I hope he dies of lung cancer Soon',
                                   'Hey you are a shithead stop vandilizing articles',
                                   'You are a bunch of wikipedia nazis stop removing information just because YOU dont know if a piece of information is true doesnt mean its not true',
                                  'that faggot kevin bravo',
                                  'I still am annoyed at you blockig me and DMacks what an idiot',
                                  'homophobic Mexican arsehole why arent you banned yet',
                                  'You are a fag are you serious Seriously what purpose does flagging my article for deletion serve',
                                  'Maybe youre just ashamed of enjoying looking at the picture and dont want it to be cp because you are a pedo',
                                  'wow So making edits to my own talk page is considered vandalism wow just wow ED has you guys dead on You really do have no lives and you guys are a bunch of pedos',
                                  'Oh I also take 1000 responsibility I feel worthless and like a piece of shit that should be lynched and killed']})
scores2=[]
for index, row in test_2_df.iterrows():
    comment = row['Comment']
    try:
      score = get_toxicity_score(comment)
    except:
      score = 0
    scores2.append(score)
   
test_2_df["toxicity score"]=scores2


score_diff=[]
for x  in range(len(scores2)):
    diff=scores2[x]-scores[x]
    score_diff.append(diff)

test_2_df['Score Change']=score_diff
  
test_2_df



                       

Comment  \
0                                                                           How dare you vandalize that page about the HMS Beagle you are a demon dont vandalize again   
1                                                                                                            No that arrogant self serving immature idiot Get it right   
2                                                                                                                                                 Its simple stupidity   
3                                  Being drunk made Gibson say what he really believes The little freak Nazi is like his father and I hope he dies of lung cancer Soon   
4                                                                                                                     Hey you are a shithead stop vandilizing articles   
5                   You are a bunch of wikipedia nazis stop removing information just because YOU dont know if a piece of information is true doesnt mean its not true   
6                                                                                                                                              that faggot kevin bravo   
7                                                                                                        I still am annoyed at you blockig me and DMacks what an idiot   
8                                                                                                                 homophobic Mexican arsehole why arent you banned yet   
9                                                                     You are a fag are you serious Seriously what purpose does flagging my article for deletion serve   
10                                                        Maybe youre just ashamed of enjoying looking at the picture and dont want it to be cp because you are a pedo   
11  wow So making edits to my own talk page is considered vandalism wow just wow ED has you guys dead on You really do have no lives and you guys are a bunch of pedos   
12                                                      Oh I also take 1000 responsibility I feel worthless and like a piece of shit that should be lynched and killed   

    toxicity score  Score Change  
0        64.912040      7.692160  
1        85.850734     -1.936286  
2        64.262060    -25.620320  
3        81.156270      0.000000  
4        89.882380      2.915300  
5        69.899110      9.644226  
6        81.156270     -5.140450  
7        85.333383      0.730653  
8        92.489950      0.864564  
9        84.031910     14.132800  
10       63.085170      4.233460  
11       78.207105      6.312805  
12       91.103500      0.000000

In [103]:
test_3_df=pd.DataFrame({'Comment1':['we hate america and we are going to bomb the shit out of there cities Quezas will rain',
                                   'I love to kill babies it my favorite thing to do in the world',
                                   'Black people should starve and die',
                                   'I want to go shoot a school full of kids up',
                                   ], 'Comment2':['You are stupid',
                                                                              'He is a dumb dumb',
                                                                             'Kevin Bravo is a gay idiot',
                                                                                    'Joel is such a pussy in the last of us']})

scores3=[]
for index, row in test_3_df.iterrows():
    comment = row['Comment1']
    try:
      score = get_toxicity_score(comment)
    except:
      score = 0
    scores3.append(score)
scores4=[]
for index, row in test_3_df.iterrows():
    comment = row['Comment2']
    try:
      score = get_toxicity_score(comment)
    except:
      score = 0
    scores4.append(score)

test_3_df["Toxicity Score1"]=scores3
test_3_df["Toxicity Score2"]=scores4

test_3_df

Comment1  \
0  we hate america and we are going to bomb the shit out of there cities Quezas will rain   
1                           I love to kill babies it my favorite thing to do in the world   
2                                                      Black people should starve and die   
3                                             I want to go shoot a school full of kids up   

                                 Comment2  Toxicity Score1  Toxicity Score2  
0                          You are stupid        85.850734        91.103500  
1                       He is a dumb dumb        84.031910        85.173553  
2              Kevin Bravo is a gay idiot        93.383175        95.637540  
3  Joel is such a pussy in the last of us        75.109370        84.031910

     My hyptothesis was that Perspective API's model is more bias towards comments using direct langauge and verbage. For example comments with phrases such as "you are" or "he is" will genrally result in higher toxicity scores than comments using less or no direct language. In order to test this hypothesis I pulled thirteen comments from the sammple data and found the toxicity score of each one. I then re wrote every comment in a new data frame. If the comment used direct language I re-wrote without direct language and if didn't use direct langauge I re-wrote using direct langauge. I then found the toxicity score of the new comments that I had written as well as calculated the difference between the new and old toxicities. 
     For the purpose of this test my threshold for what is considered a toxic comment is 50% and decided a change in toxicity of 3% was considered statistically significant. I also decided if more than half of the comments had a signifcant change in toxicity I can declare my hypotheis to be true. Out of the thirteen comments, seven of them had significant change in toxicity when direct language was either removed or added. seven out of thirteen comments gives us a 53.8% which is above my threshold therefore my hypthesis has been declared infact true. I also believe the success rate would be higher using a bigger sample size.
     After conducting my test it is pretty clear to me that after conducting this test that Perspective API is very biased towards comments using direct language. To even further prove my hypothesis true I conduced a small second test. I gathered some of the most hurtful toxic comments I could find(going by context not score)and compared them to short comments using direct langauge. For example the comment "we hate america and we are going to bomb the shit out of there cities Quezas will rain" had a toxicity score of 85% which was lower than the comment "You are stupid" which had a toxicity score of 91%. This is true for other pairs of comments I tested. This is also tells me the Persepcitve API is not great about picking up context as a comment like "I want to go shoot a school full of kids up" is really awful and toxic comment yet only has a toxicity of 75% compared to "He is a dumb dumb" which has a score of 85%. It appears that since the school shooting comment doesn't actually have any really bad individual words could be the reason why the score isn't higher. However, A.I. is becoming increasinly intelligent and APIs such as google's perspecitce will eventually catch up and grow with time. We already have A.I. systems in place like chat GPT and GPT 4 has just come out so its definitely here to stay and will only become better.
     
 